From 064e024320d7c5bd4874c00a64c9d5a4f923e647 Mon Sep 17 00:00:00 2001
From: Stefano Babic <stefano.babic@swupdate.org>
Date: Thu, 5 Oct 2023 10:20:13 +0200
Subject: [PATCH] Always link libubootenv

Library can be used to store an arbitrary set of variables and not only
the U-Boot's environment. This will be used to save application
variables and SWUpdate's state.

Signed-off-by: Stefano Babic <stefano.babic@swupdate.org>
---
 Makefile.flags     |  5 +----
 bootloader/uboot.c | 25 +++++++------------------
 2 files changed, 8 insertions(+), 22 deletions(-)

diff --git a/Makefile.flags b/Makefile.flags
index 2a021c8..cfdb7a6 100644
--- a/Makefile.flags
+++ b/Makefile.flags
@@ -106,7 +106,7 @@ export SYSROOT=$(CONFIG_SYSROOT)
 endif
 
 # Links always pthread
-LDLIBS += pthread
+LDLIBS += pthread ubootenv
 # lua
 ifneq ($(CONFIG_LUA),)
 LDFLAGS_swupdate += -Wl,-E
@@ -219,9 +219,6 @@ endif
 
 
 ifeq ($(CONFIG_BOOTLOADER_STATIC_LINKED),y)
-ifeq ($(CONFIG_UBOOT),y)
-LDLIBS += ubootenv
-endif
 ifeq ($(CONFIG_BOOTLOADER_EBG),y)
 LDLIBS += ebgenv
 endif
diff --git a/bootloader/uboot.c b/bootloader/uboot.c
index 7fd47ac..727ee0f 100644
--- a/bootloader/uboot.c
+++ b/bootloader/uboot.c
@@ -127,9 +127,15 @@ static bootloader uboot = {
 	.apply_list = &do_apply_list
 };
 
+/*
+ * libubootenv is not only used as interface to U-Boot.
+ * It is also used to save SWUpdate's persistent variables that
+ * survives after a restart of the device but should not be
+ * considered by the bootloader. That requires libubootenv
+ * is always linked.
+ */
 static bootloader* probe(void)
 {
-#if defined(BOOTLOADER_STATIC_LINKED)
 	libuboot.open = libuboot_open;
 	libuboot.close = libuboot_close;
 	libuboot.exit = libuboot_exit;
@@ -139,23 +145,6 @@ static bootloader* probe(void)
 	libuboot.load_file = libuboot_load_file;
 	libuboot.set_env = libuboot_set_env;
 	libuboot.env_store = libuboot_env_store;
-#else
-	void* handle = dlopen("libubootenv.so.0", RTLD_NOW | RTLD_GLOBAL);
-	if (!handle) {
-		return NULL;
-	}
-
-	(void)dlerror();
-	load_symbol(handle, &libuboot.open, "libuboot_open");
-	load_symbol(handle, &libuboot.close, "libuboot_close");
-	load_symbol(handle, &libuboot.exit, "libuboot_exit");
-	load_symbol(handle, &libuboot.initialize, "libuboot_initialize");
-	load_symbol(handle, &libuboot.get_env, "libuboot_get_env");
-	load_symbol(handle, &libuboot.read_config, "libuboot_read_config");
-	load_symbol(handle, &libuboot.load_file, "libuboot_load_file");
-	load_symbol(handle, &libuboot.set_env, "libuboot_set_env");
-	load_symbol(handle, &libuboot.env_store, "libuboot_env_store");
-#endif
 	return &uboot;
 }
 
